frontend/simple/views/Join.vue

```diff
@@ -97,36 +97,42 @@
     </section>
 </template>
 <script>
+import {namespace} from '../js/backend/hapi'
 import * as Events from '../../../shared/events'
 import backend from '../js/backend/'
-import * as db from '../js/database'
-import {latestContractState} from '../js/state'
+import { latestContractState } from '../js/state'
 import L from '../js/translations'
+
 export default {
   name: 'Join',
-  mounted () {
-    this.fetchData()
+  async mounted () {
+    let state = await latestContractState(this.$route.query.groupId)
+    console.log(state)
+    if (!state.invitees.find(invitee => invitee === this.$store.state.loggedIn.name)) {
+      // TODO: proper user-facing error
+      console.log(new Error('Invalid Invitation'))
+      this.$router.push({path: '/mailbox'})
+    }
+    for (var i = 0; i < state.members.length; i++) {
+      const contractId = await namespace.lookup(state.members[i])
+      state.members[i] = await latestContractState(contractId)
+    }
+    this.contract = state
   },
   methods: {
-    fetchData: async function () {
-      let state = await latestContractState(this.$route.query.groupId)
-      if (!state.invitees.find(invitee => invitee === this.$store.state.loggedIn)) {
-        console.log(new Error('Invalid Invitation'))
-        await this.$store.dispatch('deleteMail', this.$route.query.inviteHash)
-        this.$router.push({path: '/mailbox'})
-      }
-      this.contract = state
-    },
     accept: async function () {
       try {
         this.errorMsg = null
+        let latest = await backend.latestHash(this.$route.query.groupId)
+        let acceptance = new Events.HashableGroupAcceptInvitation({ username: this.$store.state.loggedIn.name, inviteHash: this.$route.query.inviteHash, acceptanceDate: new Date() }, latest)
+        this.$store.commit('setCurrentGroupId', this.$route.query.groupId)
         await backend.subscribe(this.$route.query.groupId)
         await this.$store.dispatch('syncContractWithServer', this.$route.query.groupId)
-        let latest = await db.recentHash(this.$route.query.groupId)
-        let acceptance = new Events.AcceptInvitation({ username: this.$store.state.loggedIn, acceptanceDate: new Date() }, latest)
-        this.$store.commit('setCurrentGroupId', this.$route.query.groupId)
         await backend.publishLogEntry(this.$route.query.groupId, acceptance)
-        this.$store.dispatch('deleteMessage', this.$route.query.inviteId)
+        let identityContractId = await namespace.lookup(this.contract.founderUsername)
+        await backend.subscribe(identityContractId)
+        await this.$store.dispatch('syncContractWithServer', identityContractId)
+        this.$store.commit('deleteMessage', this.$route.query.inviteHash)
```

Comment:
Also, (replying here because otherwise comment will appear on the left hand side of the screen), for the `catch` below, is it really gracefully handling all failure modes?

Each `await` can result in a failure and a different application state. We don't want to end up in a weird application state after a failure. If any "cleanup" needs to be done to "revert" us to the previous state then it should be done in the catch. Kinda like a database transaction.